proxy/src/telemetry/sensor/http.rs

```diff
@@ -434,12 +434,23 @@ where
 
     fn poll_data(&mut self) -> Poll<Option<Self::Data>, h2::Error> {
         let frame = try_ready!(self.sense_err(|b| b.poll_data()));
+        let is_eos = self.is_end_stream();
         let frame = frame.map(|frame| {
             let frame = frame.into_buf();
-            if let Some(ref mut inner) = self.inner {
+            self.inner = self.inner.take().and_then(|mut inner| {
                 *inner.frames_sent() += 1;
                 *inner.bytes_sent() += frame.remaining() as u64;
-            }
+                if is_eos {
```

Comment:
this seems slightly more complicated than i'd expect.

Can't we just use the value-ness of `frame` to know whether we're at EOS?  I.e. we can't actualyl return _data_ and be at EOS -- this has to return None to indicate that.

```rust

    fn poll_data(&mut self) -> Poll<Option<Self::Data>, h2::Error> {
        let frame = try_ready!(self.sense_err(|b| b.poll_data())).map(|f| f.into_buf());

        match frame {
            None => {
                if let Some(inner) = self.inner.take() {
                    inner.end(None);
                }
            }
            Some(ref f) => {
                if let Some(ref mut inner) = self.inner {
                    *inner.frames_sent() += 1;
                    *inner.bytes_sent() += f.remaining() as u64;
                }
            }
        }

        Ok(Async::Ready(frame))
    }
```

Is this equivalent? I find this logic to be a little clearer:
* we don't do any sort of side-effecty logic in `map`/`and_then`
* we only mutate `inner` when it's absolutely necessary
* the reader doesn't have to worry about the semantics of `is_end_stream` versus the underlying type.

Reply:
Yeah, I think something like that is equivalent (and certainly cleaner!). I'll get back to you once I've got some real tests for this.

Reply:
@olix0r I tried this, and it does _not_ appear to fix the failing telemetry tests.

I think it's necessary to send the `StreamResponseEnd` event as soon as `self.body.is_end_stream()` is true. It looks like`tower_h2::Flush` does not appear to call `poll_data` again after we see the EOS DATA frame, so we never see `None` here.

Reply:
So, this *will* need to check `is_end_stream`, but it seems preferable to only pop out of `inner.take()` if it is the end. So, sort of a mix of the proposed changes and the ones mentioned by @olix0r above.

Reply:
Hopefully 9ec9f32c154ed4ed0a79e4938f89c716cf9b448b makes this somewhat clearer!